### PR TITLE
Add IVP function support and data assimilation examples

### LINKED ISSUE #494: Add a data assimilation problem
Add some data assimilation problems (Culham? Reading?)

### PR DESCRIPTION
**Pull request by AndrewLister-STFC**
#### Description of Work

Adds support for parsing IVP problems.

The IVP problems are defined by python functions which look like:
```
def func(t, x, *p):
    ...
    return np.array([...])
```
These are used to define the RHS of the vector equation:
```
x' = f(t, x, *p)
```

Fixes #494

Caveats:
- This is slow. It solves an IVP for every function evaluation...
- This only works with plots off

TODO:
- [x] Add more examples - noisy, different system
- [x] Make plot handling work / fail gracefully
- [x] Docs (once function creation is unlikely to be altered)
- [x] Tests

#### Testing Instructions

1. Try running the data assimilation examples!
2. Look at the format for new problems - does it seem possible to add new IVP/data assimilation problems?
3. Proof read the docs

Function: Does the change do what it's supposed to?

Tests: Does it pass? Is there adequate coverage for new code?

Style: Is the coding style consistent? Is anything overly confusing?

Documentation: Is there a suitable change to documentation for this change?

### CODE REVIEW

**Comment by AndrewLister-STFC** [comment]
@tyronerees Do you know why numdifftools seems to convert the params to a complex space?

**Review by tyronerees — CHANGES_REQUESTED** [review]
This looks great -- there's tons of potential here for people to working in data assimilation to use this!

One thing I do think is missing is some more docs -- there's not much to explain what's going on.  Could you add something to the page [here](https://fitbenchmarking--779.org.readthedocs.build/en/779/users/BenchmarkProblems.html) describing what's actually happening.

Some ideas, but probably for other pull requests:
- it would be good to include analytic Jacobians
- we should add plots
- (related to the plots) data assimilation is all about forecasting, so it would be good to add a parameter to plot the forecast window (basically, you expect the learned parameters to fit the data you have well, but that may, or may not, predict the future well)

We can discuss and create issues for these together.

**Review by tyronerees — CHANGES_REQUESTED** [review]
Forgot to run pylint -- I also get:

```
************* Module lorentz
examples/benchmark_problems/Data_Assimilation/functions/lorentz.py:1:0: C0114: Missing module docstring (missing-module-docstring)
examples/benchmark_problems/Data_Assimilation/functions/lorentz.py:19:4: R1705: Unnecessary "elif" after "return" (no-else-return)
examples/benchmark_problems/Data_Assimilation/functions/lorentz.py:4:14: W0613: Unused argument 't' (unused-argument)
************* Module simplified_anac
examples/benchmark_problems/Data_Assimilation/functions/simplified_anac.py:1:0: C0114: Missing module docstring (missing-module-docstring)
examples/benchmark_problems/Data_Assimilation/functions/simplified_anac.py:17:4: R1705: Unnecessary "elif" after "return" (no-else-return)
examples/benchmark_problems/Data_Assimilation/functions/simplified_anac.py:4:20: W0613: Unused argument 't' (unused-argument)
************* Module benchmark_problems.scripts.generate_data_assim
examples/benchmark_problems/scripts/generate_data_assim.py:9:0: C0116: Missing function or method docstring (missing-function-docstring)
************* Module fitbenchmarking.parsing.fitbenchmark_parser
fitbenchmarking/parsing/fitbenchmark_parser.py:71:8: R1720: Unnecessary "elif" after "raise" (no-else-raise)
fitbenchmarking/parsing/fitbenchmark_parser.py:116:16: W0622: Redefining built-in 'input' (redefined-builtin)
fitbenchmarking/parsing/fitbenchmark_parser.py:59:4: R0915: Too many statements (76/60) (too-many-statements)
fitbenchmarking/parsing/fitbenchmark_parser.py:397:0: R1721: Unnecessary use of a comprehension (unnecessary-comprehension)
fitbenchmarking/parsing/fitbenchmark_parser.py:387:8: W0201: Attribute '_mantid_starting_values' defined outside __init__ (attribute-defined-outside-init)
fitbenchmarking/parsing/fitbenchmark_parser.py:455:8: W0201: Attribute '_ivp_equation_name' defined outside __init__ (attribute-defined-outside-init)
fitbenchmarking/parsing/fitbenchmark_parser.py:456:8: W0201: Attribute '_ivp_starting_values' defined outside __init__ (attribute-defined-outside-init)
```

**Comment by AndrewLister-STFC** [comment]
> 
> 
> This looks great -- there's tons of potential here for people to working in data assimilation to use this!
> 
> One thing I do think is missing is some more docs -- there's not much to explain what's going on. Could you add something to the page [here](https://fitbenchmarking--779.org.readthedocs.build/en/779/users/BenchmarkProblems.html) describing what's actually happening.
> 
> Some ideas, but probably for other pull requests:
> 
>     * it would be good to include analytic Jacobians
> 
>     * we should add plots
> 
>     * (related to the plots) data assimilation is all about forecasting, so it would be good to add a parameter to plot the forecast window (basically, you expect the learned parameters to fit the data you have well, but that may, or may not, predict the future well)
> 
> 
> We can discuss and create issues for these together.

I forgot this page was even there! I guess I also need to package up the examples?

**Comment by AndrewLister-STFC** [comment]
My understanding is that the issues to split off are:
- Add time series data
- Add a way of plotting the time series data with the prediction. Possibly splitting the data so that there's something to compare the prediction to.
- Add analytic jacobians

File: docs/source/users/BenchmarkProblems.rst [hunk 1]
```diff
@@ -199,6 +199,23 @@ that are added, and also that any new parsers reimplement these
...
 
+Data Assimilation
+=================
+
+**Download** :download:`.zip <http://numerical.rl.ac.uk/fitbenchmarking/data_assimilation.zip>`
+or :download:`.tar.gz <http://numerical.rl.ac.uk/fitbenchmarking/data_assimilation.tar.gz>`
+
+This folder (also found in `examples/benchmark_problems/data_assimilation`) contains
+two examples using the data assimilation problem definition in fitbenchmarking.
+These example follow the method set out in
```
Comment (tyronerees) [inline docs/source/users/BenchmarkProblems.rst:210]:
```suggestion
These examples follow the method set out in 
```

**Review by tyronerees — CHANGES_REQUESTED** [review]
Thanks @AndrewLister-STFC -- just one typo here, then all good.  Sorry for the long review period!

**Review by tyronerees — APPROVED** [review]

**Pull request merged**